teps/0086-changing-the-way-result-parameters-are-stored.md

```diff
@@ -0,0 +1,444 @@
+file name, for example, "/teps/images/NNNN-workflow.jpg".
+-->
+
+### Result References
+
+Beyond size limits of specific TaskRuns’ results, the fundamental issue is that API objects in etcd are not suitable for storing as much data as users want to be able to report in results, and pass as parameters.
+
+In addition to disaggregating TaskRun statuses into PipelineRun statuses (which we should do anyway before v1), we should introduce some way to pass one TaskRun’s output results into another TaskRun’s input parameters by reference, instead of only by value as they are today.
+
+**Example:**
+
+```yaml
+tasks:
+- name: first-task
+  taskSpec: foo
+  results:
+  - name: out
+    byRef: true # <-- *new field*
+- name: second-task
+  taskSpec: bar
+  params:
+  - name: in
+    value: "$(tasks.first-task.results.out)"
+```
+
+
+**Implementation:**
+
+Result references require some code to take the contents of `/tekton/results/out` in the container environment, copy them elsewhere, and mint a unique reference. It also requires some code to take that reference, locate the contents whever they are, and make them available to other TaskRuns at `/tekton/params/in`.
+
+Both of these could be satisfied by the same component: a Tekton-internal sidecar, the *Result Collector Sidecar*. 
+
+**Sidecar API:**
+
+1. The controller is passed a flag, -sidecar-image=<image>
+2. This image is added as a container in every TaskRun Pod, with each steps’ results emptyDir volume attached as read–only, and params emptyDir volumes attached as read-write.
+3. Watch directory, and any time it sees a result file appear, it writes that content externally (GCS, datastore, cuneiform tablets, the moon)
+it must be able to produce an opaque string that represents that result value, which it writes to TaskRun status
+> TODO: Do we achieve this via terminationMessage? We can enforce that these opaque strings be small; otherwise, ConfigMap? – an example signed URL is 824 bytes, this limits to ~12 per TaskRun.
+4. the TaskRun controller updates the status to include that opaque string
+5. when the next TaskRun starts, for each passed-in param that’s passed by reference, the TaskRun controller passes opaque reference strings to the sidecar along with the associated param name. the sidecar dereferences the opaque strings and writes the real param value to the param value file mounted into steps.
+6. the sidecar reports READY only after these steps are complete, allowing the first step to begin as normal.
+
+```yaml 
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+...
+status:
+  taskResults:
+  - name: out
+    valueRef: <opaque-string>  # <-- *new field*
+```
+
+With this API in place, Tekton can provide an implementation that writes results to a ConfigMap, and other operators can implement their own that write results to cluster-local persistent volumes, external object storage, document store, relational database, etc.
```

Comment:
If we write the results to a ConfigMap by default, I think this means by default the controller needs to ship with permissions to modify service account roles and bindings - and I don't think that's a good default for us to have (and we know a number of users won't be able to use this).

I think we need to find a default that allows us to standardize on a result size but also doesn't require these extra permissions - even if that option has some other downsides.